include/dgl/immutable_graph.h

```diff
@@ -489,39 +603,51 @@ class ImmutableGraph: public GraphInterface {
    *       first, and ties are broken by the order of edge ID.
    * \return EdgeArray containing all edges between all pairs.
    */
-  EdgeArray EdgeIds(IdArray src, IdArray dst) const;
+  EdgeArray EdgeIds(IdArray src, IdArray dst) const override {
+    if (in_csr_) {
+      return in_csr_->EdgeIds(dst, src);
+    } else {
+      return GetOutCSR()->EdgeIds(src, dst);
+    }
+  }
 
   /*!
    * \brief Find the edge ID and return the pair of endpoints
    * \param eid The edge ID
    * \return a pair whose first element is the source and the second the destination.
    */
-  std::pair<dgl_id_t, dgl_id_t> FindEdge(dgl_id_t eid) const;
+  std::pair<dgl_id_t, dgl_id_t> FindEdge(dgl_id_t eid) const override {
+    return GetCOO()->FindEdge(eid);
+  }
 
   /*!
    * \brief Find the edge IDs and return their source and target node IDs.
    * \param eids The edge ID array.
    * \return EdgeArray containing all edges with id in eid.  The order is preserved.
    */
-  EdgeArray FindEdges(IdArray eids) const;
+  EdgeArray FindEdges(IdArray eids) const override {
+    return GetCOO()->FindEdges(eids);
+  }
 
   /*!
    * \brief Get the in edges of the vertex.
    * \note The returned dst id array is filled with vid.
    * \param vid The vertex id.
    * \return the edges
    */
-  EdgeArray InEdges(dgl_id_t vid) const {
-    return this->GetInCSR()->GetEdges(vid);
+  EdgeArray InEdges(dgl_id_t vid) const override {
+    const EdgeArray& ret = GetInCSR()->OutEdges(vid);
```

Comment:
a reference to a return value. This is very dangerous.

Reply:
This is totally fine for handling readonly temporary object.

Reply:
I see. I didn't know that.
https://stackoverflow.com/questions/2822243/store-return-value-of-function-in-reference-c

Reply:
is this allowed from the very beginning of C++ specification? This semantic looks really weird to me.